node/gum/src/lib.rs

```diff
@@ -112,7 +113,55 @@ pub use jaeger::hash_to_trace_identifier;
 #[doc(hidden)]
 pub use polkadot_primitives::{CandidateHash, Hash};
 
-pub use gum_proc_macro::{debug, error, info, trace, warn};
+pub use gum_proc_macro::{debug, error, info, trace, warn, warn_if_frequent};
 
 #[cfg(test)]
 mod tests;
+
+/// Max allowed frequency of debug messages
+pub const MAX_FREQ_RATE: f64 = 1.0;
+const MAX_FREQ_SIZE: usize = 10;
+
+/// Utility struct to compare the rate of its own calls.
+pub struct Freq {
+	timestamps: VecDeque<u64>,
```

Comment:
I think it is an overkill for these purposes. We really don't need any data structure, especially `VecDequeue` for this task as we can use a simple and quite straightforward approach here: just use an [exponential moving average](https://en.wikipedia.org/wiki/Exponential_smoothing). In this case, we would need to store just 3 values: the last message time, the current average and the samples count. With alpha parameter equal to `0.5` we will expire old values quite quickly and measure the average quite efficiently in all the cases. Here is a code for that:

```rust
struct EmaBucket {
  count: u32,
  current: f32,
}

fn update_bucket(bucket: &mut EmaBucket, value: f32, alpha: f32) -> f32 {
	let diff = value - bucket.current;
	let incr = diff * alpha;
  bucket.current += incr;
	bucket.count += 1;

	return bucket.current;
}
```

I have not included the last message time here, but it should be obvious how to use it as our `value` is actually `now() - last`.